source/module_hsolver/hsolver_pw.cpp

```diff
@@ -572,7 +574,9 @@ void HSolverPW<T, Device>::hamiltSolvePsiK(hamilt::Hamilt<T, Device>* hm,
             ModuleBase::timer::tick("David", "spsi_func");
         };
 
-        DiagoDavid<T, Device> david(pre_condition.data(),
+        // const auto pre_op = make_pre_op(pre_condition, fvec::div_prevec<T, Device>);
+        const PreOP<T, Device> pre_op(pre_condition);
+        DiagoDavid<T, Device> david(pre_op.get(),
                                     nband,
                                     dim,
                                     PARAM.inp.pw_diag_ndim,
```

Comment:
besides the `PreOP`, now I realize the `DiagoDavid` class has the similar problem. It is confusing to instantiate the class `DiagoDavid` then name as `david`. A factory function will be better and more consistent, but will not be work of this PR:
```c++
std::unique_ptr<Diago> dav = make_diago("dav", /* other args */); // or use object directly without pointer
```